baseapp/abci.go

```diff
@@ -520,9 +530,25 @@ func (app *BaseApp) ProcessProposal(_ context.Context, req *abci.RequestProcessP
 		return nil, errors.New("ProcessProposal called with invalid height")
 	}
 
-	// always reset state given that ProcessProposal can timeout and be called again
-	emptyHeader := cmtproto.Header{ChainID: app.chainID}
-	app.setState(runTxProcessProposal, emptyHeader)
+	// Always reset state given that ProcessProposal can timeout and be called
+	// again in a subsequent round.
+	header := cmtproto.Header{
+		ChainID:            app.chainID,
+		Height:             req.Height,
+		Time:               req.Time,
+		ProposerAddress:    req.ProposerAddress,
+		NextValidatorsHash: req.NextValidatorsHash,
+	}
+	app.setState(execModeProcessProposal, header)
+
+	// Since the application can get access to FinalizeBlock state and write to it,
+	// we must be sure to reset it in case ProcessProposal timeouts and is called
+	// again in a subsequent round. However, we only want to do this after we've
+	// processed the first block, as we want to avoid overwriting the finalizeState
+	// after state changes during InitChain.
+	if req.Height > app.initialHeight {
+		app.setState(execModeFinalize, header)
+	}
```

Comment:
> So it's not possible that ProcessProposal gets called more than once during the first block?

It is possible. Why does that matter? This conditional is irrelevant to the round number in the first block.

Reply:
Then I'm not so sure I understood you the other day lol. In other words, why this wouldn't cause an issue? I understand that we would be having multiple writes to state without discarding what was done in the previous processProposal.

```
1. InitChain writes to FinalizeState
2. ProcessProposal writes to FinalizeState
3. ProcessProposal gets called again and writes to FinalizeState
```

So step 3 would be seeing the results of steps 1+2, but I think it should only be seeing results of step 1.

Reply:
That's not the full correct understanding of what's happening. What is actually happening is, _in the context of the first block_:

```
1. InitChain writes to FinalizeState
2. ProcessProposal writes to a CACHED copy of FinalizeState
3. If ProcessProposal from (2) times out, (2) is called again
```

So step (3) would be a on a fresh cached copy of `FinalizeState` from step (1).

The big caveat here though is that a chain should NOT write to `FinalizeState` on the first block. I'll update the godoc. This isn't a concern really though because vote extensions can be used until at least the block after the 1st block.

Reply:
Ahh thanks for taking the time to explain me, I was a bit too worried 😅